### PR TITLE
patch service account in credential api

### PR DESCRIPTION
**Pull request by akihikokuroda**
This is replacement of PR https://github.com/tektoncd/dashboard/pull/59.  I made the same changes in th clean branch.

### CODE REVIEW

**Comment by akihikokuroda** [comment]
This PR is for issue #58

File: pkg/endpoints/credentials.go [hunk 1]
```diff
@@ -153,6 +156,20 @@ func (r Resource) createCredential(request *restful.Request, response *restful.R
...
+	// Patch service account
+	saName := ""
+	providedSaName := os.Getenv("PIPELINE_RUN_SERVICE_ACCOUNT")
+	if providedSaName != "" {
+		saName = providedSaName
+	} else {
+		saName = "default"
+	}
+	if !r.addSecretToSA(saName, cred.Name, requestNamespace) {
+		errorMessage := fmt.Sprintf("Error adding secret in service account: %s", saName)
```
Comment (a-roberts) [inline pkg/endpoints/credentials.go]: Should be lower case e for the error message string

File: Gopkg.lock [hunk 2]
```diff
@@ -158,12 +158,12 @@
...
 [[projects]]
   branch = "master"
-  digest = "1:b6fca2d4933db5050e8729d0931acc650c2492daa13decc64ad62d48145f7d20"
+  digest = "1:58dc2abfa0d3901dde142cd636abf13d495f37c125564390c7a5791d1fa29e80"
   name = "github.com/knative/pkg"
   packages = [
     "apis",
     "apis/duck",
-    "apis/duck/v1beta1",
+    "apis/duck/v1alpha1",
```
Comment (a-roberts) [inline Gopkg.lock]: I'd rather this be v1beta1 than alpha

File: pkg/endpoints/credentials.go [hunk 3]
```diff
@@ -345,3 +371,82 @@ func credentialToSecret(cred credential, namespace string, response *restful.Res
 	// Return secret
 	return &secret, true
 }
+
+// addSecretToSA - Add secret to service account 
+func (r Resource) addSecretToSA(saName string, secretName string , namespaceName string) bool {
+	sa := corev1.ServiceAccount{
```
Comment (a-roberts) [inline pkg/endpoints/credentials.go]: Looks off to me, and the next line (whitespace) I'd rather this be
```
func (r Resource) addSecretToSA(saName, secretName, namespaceName string) bool {
  a := corev1.ServiceAccount{
	  Secrets: []corev1.ObjectReference {
      {
        Name: secretName
      },
  },
}
```

File: pkg/endpoints/credentials.go [hunk 4]
```diff
@@ -345,3 +371,82 @@ func credentialToSecret(cred credential, namespace string, response *restful.Res
...
+		Secrets: []corev1.ObjectReference {{Name: secretName},},
+	}
+
+	marshaledSA, err := json.Marshal(sa)
+	if err != nil {
+		logging.Log.Errorf("failed to marshal payload")
```
Comment (a-roberts) [inline pkg/endpoints/credentials.go]: Mention it's failed to `marshal` the payload for the service account

File: pkg/endpoints/credentials.go [hunk 5]
```diff
@@ -345,3 +371,82 @@ func credentialToSecret(cred credential, namespace string, response *restful.Res
...
+		return false
+	}
+
+	logging.Log.Debugf("Service account to patch is %s", saName)
```
Comment (a-roberts) [inline pkg/endpoints/credentials.go]: These first two `debugf` statements should be at the top of this method before the call to `json.Marshal`

File: pkg/endpoints/credentials.go [hunk 6]
```diff
@@ -345,3 +371,82 @@ func credentialToSecret(cred credential, namespace string, response *restful.Res
...
+			found = true
+			entry = i
+			break
+		}
+	}
+	if found {
+		path := fmt.Sprintf("/secrets/%d", entry)
+		data := []Patch {{ Operation: "remove", Path: path }} 
+		patch, err := json.Marshal(data)
+		logging.Log.Debugf("Patch JSON:%s", string(patch))
```
Comment (a-roberts) [inline pkg/endpoints/credentials.go]: Add a space before the `%s` here

File: pkg/endpoints/credentials.go [hunk 7]
```diff
@@ -153,6 +156,20 @@ func (r Resource) createCredential(request *restful.Request, response *restful.R
 		return
 	}
 
+	// Patch service account
+	saName := ""
+	providedSaName := os.Getenv("PIPELINE_RUN_SERVICE_ACCOUNT")
+	if providedSaName != "" {
+		saName = providedSaName
```
Comment (a-roberts) [inline pkg/endpoints/credentials.go]: PIPELINE_RUN_SERVICE_ACCOUNT isn't anywhere in the code atm (e.g. in https://github.com/tektoncd/dashboard/blob/master/config/tekton-dashboard-deployment.yaml), so actually it's worth doing the following

1) omit this altogether and document in the readme the dashboard must use the default service account
2) add the variable into our deployment yaml and test all is good with and without the service account being set

File: pkg/endpoints/credentials_test.go [hunk 8]
```diff
@@ -501,3 +507,19 @@ func (r Resource) getK8sCredential(namespace string, name string) (credential cr
 	}
 	return secretToCredential(secret,false)
 }
+
+// Get check service account secret
+func (r Resource) checkSecretInSa(namespace string, nameSa string, nameSecret string, t *testing.T) bool {
```
Comment (a-roberts) [inline pkg/endpoints/credentials_test.go]: Formatting off here, should be using tabs (tab size 2)

File: pkg/endpoints/credentials.go (hunk 1 above)
Comment (akihikokuroda) [reply pkg/endpoints/credentials.go]: I fix it.

File: Gopkg.lock (hunk 2 above)
Comment (akihikokuroda) [reply Gopkg.lock]: I fix it.

File: pkg/endpoints/credentials.go (hunk 3 above)
Comment (akihikokuroda) [reply pkg/endpoints/credentials.go]: I fix it.

File: pkg/endpoints/credentials.go (hunk 4 above)
Comment (akihikokuroda) [reply pkg/endpoints/credentials.go]: I fix it.

File: pkg/endpoints/credentials.go (hunk 5 above)
Comment (akihikokuroda) [reply pkg/endpoints/credentials.go]: I fix it.

File: pkg/endpoints/credentials.go (hunk 6 above)
Comment (akihikokuroda) [reply pkg/endpoints/credentials.go]: I fix it.

File: pkg/endpoints/credentials.go (hunk 7 above)
Comment (akihikokuroda) [reply pkg/endpoints/credentials.go]: I did the "2".

File: pkg/endpoints/credentials_test.go (hunk 8 above)
Comment (akihikokuroda) [reply pkg/endpoints/credentials_test.go]: I can not see the issue.  The format looks OK for me.

Comment (a-roberts) [reply pkg/endpoints/credentials_test.go]: Was referring to the entire method, but I'll take it!

**Pull request merged**